pycloudlib/ibm/instance.py

```diff
@@ -479,35 +481,132 @@ def __init__(
+            client: IBM VPC client
+            instance: raw instance dict
+            username: username to use for SSH
+            kwargs: other keyword arguments
+        """
         return cls(
             *args,
             client=client,
             instance=instance,
-            floating_ip=floating_ip,
             username=username,
             **kwargs,
         )
 
+    def _choose_from_existing_floating_ips(
+        self,
+        zone: str,
+        name_includes="default-floating-ip",
+    ) -> dict:
+        """
+        Choose a random floating IP from existing floating IPs via substring match.
+
+        Args:
+            name_includes: string, name to filter floating IPs by
+
+        Returns:
+            A floating IP object
+
+        Raises:
+            IBMException: If no floating IPs are found matching the substring
+            IBMCapacityException: If all floating IPs matching the substring are already in use
+        """
+        floating_ips = list(
+            _iter_resources(
+                self._client.list_floating_ips,
+                resource_name="floating_ips",
+                # Select that match the desired name
+                filter_fn=lambda ip: name_includes in ip["name"]
+                and ip.get("zone")
+                and ip["zone"]["name"] == zone,
+            )
+        )
+        if not floating_ips:
+            raise IBMException(
+                f"No floating IPs found matching substring {name_includes}"
+            )
+        # filter out floating ips that are already associated with an instance
+        floating_ips = [ip for ip in floating_ips if "target" not in ip]
+        if not floating_ips:
+            raise IBMCapacityException(
+                f"All floating IPs matching substring {name_includes}",
+                "are already in use.",
+            )
+
+        # pick random floating ip to help reduce contention
+        floating_ip = random.choice(floating_ips)
+
+        self._log.info(
+            "Using existing floating ip '%s' with address %s",
+            floating_ip["name"],
+            floating_ip["address"],
+        )
+        return floating_ip
+
+    def attach_floating_ip(self, floating_ip: dict) -> Optional[dict]:
+        """
+        Attach a floating IP to this instance.
+
+        Args:
+            floating_ip: floating IP dict
+
+        Returns:
+            A floating IP dict if successfully attached, else None
+        """
+        nic_id = self._instance["primary_network_interface"]["id"]
+        self._ibm_instance_type.add_instance_network_interface_floating_ip(
+            self._client,
+            id=floating_ip["id"],
+            instance_id=self._instance["id"],
+            network_interface_id=nic_id,
+        ).get_result()
+        # sleep 1s to let cloud update then check to make sure that floating IP successfully attached
+        time.sleep(1)
+        return self._discover_floating_ip(self._client, self._instance)
```

Comment:
The dict info is not used, can we simplify it and return a boolean representing success?

Reply:
hmmmm i kinda like the idea of returning the dict but you are right it is not used and doesn't really add much. will convert to returning a bool :saluting_face: